File: python/leetcode/791-custom-sort-string.py
class Solution:
    def customSortString(self, order: str, original: str) -> str:
        hash = {}
        for i, c in enumerate(order):
            if c not in hash:
                hash[c] = i

        for s in original:
            if s not in hash:
                hash[s] = -1

        res = sorted(original, key=lambda x: hash[x])
        return "".join(res)

    def customSortStringFast(self, S, T):
        return ''.join(sorted(T, key=S.find))

File: c#/leetcode/791-custom-sort-string.cs
public class Solution
{
  public string CustomSortString(string S, string T)
  {
    Dictionary<char, int> order = new Dictionary<char, int>();
    for (int i = 0; i < S.Length; i++)
    {
      order.Add(S[i], i);
    }

    for (int i = 0; i < T.Length; i++)
    {
      if (!order.ContainsKey(T[i]))
      {
        order.Add(T[i], -1);
      }
    }

    char[] result = T.ToArray();
    Array.Sort(result, (x, y) => order[x] - order[y]);

    return new String(result);
  }
}

